backend/llm/openai.py

```diff
@@ -2,136 +2,119 @@
-        :return: BrainPicking instance
+        :return: OpenAIBrainPicking instance
         """
         super().__init__(
-            model=model,
+            model_name=model_name,
             brain_id=brain_id,
             chat_id=chat_id,
             max_tokens=max_tokens,
             temperature=temperature,
             user_openai_api_key=user_openai_api_key,
+            streaming=streaming,
         )
-        # If user provided an API key, update the settings
-        if user_openai_api_key is not None:
-            self.settings.openai_api_key = user_openai_api_key
 
+        # Set the class attributes
         self.temperature = temperature
-        self.embeddings = OpenAIEmbeddings(openai_api_key=self.settings.openai_api_key)
-        self.supabase_client = create_client(
-            self.settings.supabase_url, self.settings.supabase_service_key
+        self.model_name = model_name
+        self.chat_id = chat_id
+        self.brain_id = brain_id
+        self.max_tokens = max_tokens
+
+    @property
+    def embeddings(self) -> OpenAIEmbeddings:
+        return OpenAIEmbeddings(openai_api_key=self.openai_api_key)
+
+    @property
+    def supabase_client(self) -> Client:
+        return create_client(
+            self.brain_settings.supabase_url, self.brain_settings.supabase_service_key
         )
-        self.llm_name = model
-        self.vector_store = CustomSupabaseVectorStore(
+
+    @property
+    def vector_store(self) -> CustomSupabaseVectorStore:
+        return CustomSupabaseVectorStore(
             self.supabase_client,
             self.embeddings,
             table_name="vectors",
-            brain_id=brain_id,
+            brain_id=self.brain_id,
         )
 
-        self.question_llm = self._create_llm(
-            model_name=self.llm_name,
-            streaming=False,
-        )
-        self.question_generator = LLMChain(
-            llm=self.question_llm, prompt=CONDENSE_QUESTION_PROMPT
-        )
+    @property
+    def question_llm(self) -> LLM:
+        return self._create_llm(streaming=False)
 
-        if streaming:
-            self.callback = AsyncIteratorCallbackHandler()
-            self.doc_llm = self._create_llm(
-                model_name=self.llm_name,
-                streaming=streaming,
-                callbacks=[self.callback],
-            )
-            self.doc_chain = load_qa_chain(
-                llm=self.doc_llm,
-                chain_type="stuff",
-            )
-            self.streaming = streaming
-        else:
-            self.doc_llm = self._create_llm(
-                model_name=self.llm_name,
-                streaming=streaming,
-            )
-            self.doc_chain = load_qa_chain(llm=self.doc_llm, chain_type="stuff")
-            self.streaming = streaming
+    @property
+    def doc_llm(self) -> LLM:
+        return self._create_llm(streaming=self.streaming, callbacks=self.callbacks)
```

Comment:
Does this 'callbacks' attribute refer to the 'callbacks' used for the streaming ?

Reply:
Yes exactly. You can also use them for a bunch of other things and I am sure we will in the future hence the seperate property decorator.